power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -273,41 +284,551 @@ inline auto rank_transformers(State const& state) -> RankedTransformerGroups {
+    using ComponentContainer = typename State::ComponentContainer;
+    using RegulatedTransformer = TapRegulatorRef<TransformerTypes...>;
+    using UpdateBuffer = std::tuple<std::vector<typename TransformerTypes::UpdateType>...>;
+
+    template <transformer_c T>
+    static constexpr auto transformer_index_of = container_impl::get_cls_pos_v<T, TransformerTypes...>;
+    static_assert(((transformer_index_of<TransformerTypes> < sizeof...(TransformerTypes)) && ...));
 
-    TapPositionOptimizer(Calculator calculator, StateUpdater updater, OptimizerStrategy strategy)
+  public:
+    TapPositionOptimizerImpl(Calculator calculator, StateUpdater updater, OptimizerStrategy strategy)
         : calculate_{std::move(calculator)}, update_{std::move(updater)}, strategy_{strategy} {}
 
-    auto optimize(State const& state) -> ResultType final {
-        auto const order = rank_transformers(state);
-        return optimize(state, order);
+    auto optimize(State const& state, CalculationMethod method) -> ResultType final {
+        auto const order = regulator_mapping<TransformerTypes...>(state, TransformerRanker{}(state));
+
+        auto const cache = this->cache_states(order);
+        auto optimal_output = optimize(state, order, method);
+        update_state(cache);
+
+        return optimal_output;
     }
 
-    constexpr auto get_strategy() { return strategy_; }
+    constexpr auto get_strategy() const { return strategy_; }
 
   private:
-    auto optimize(State const& /*state*/, RankedTransformerGroups const& /*order*/) -> ResultType {
-        // TODO(mgovers): implement outter loop tap changer
-        throw PowerGridError{};
+    auto optimize(State const& state, std::vector<std::vector<RegulatedTransformer>> const& regulator_order,
+                  CalculationMethod method) const -> ResultType {
+        initialize(regulator_order);
+
+        if (auto result = try_calculation_with_regulation(state, regulator_order, method);
+            strategy_ == OptimizerStrategy::any) {
+            return result;
+        }
+
+        // refine solution
+        step_all(regulator_order);
+        return try_calculation_with_regulation(state, regulator_order, method);
+    }
+
+    auto try_calculation_with_regulation(State const& state,
+                                         std::vector<std::vector<RegulatedTransformer>> const& regulator_order,
+                                         CalculationMethod method) const -> ResultType {
+        auto result = calculate_with_retry(state, method);
+
+        bool tap_changed = true;
+        while (tap_changed) {
+            tap_changed = false;
+            UpdateBuffer update_data;
+
+            for (auto const& same_rank_regulators : regulator_order) {
+                for (auto const& regulator : same_rank_regulators) {
+                    if (regulator.transformer.connected_at_tap_side() &&
+                        regulator.transformer.connected_at_control_side(regulator.regulator)) {
```

Comment:
I believe this check is also redundant

Reply:
you're right. i'll make an assert for this